File: Fasm.NET/Fasm.NET.cpp
/*
* Fasm.NET
* https://github.com/ZenLulz/Fasm.NET
*
* Copyright 2013 ZenLulz ~ Jämes Ménétrey
* Released under the MIT license
*
* Date: 2013-07-10
*/

#include "stdafx.h"
#include "Fasm.NET.h"

namespace Binarysharp {
	namespace Assemblers {
		namespace Fasm {
			#pragma region Static Methods
			#pragma region FasmNet::GetVersion
			/// <summary>
			/// Gets the version of FASM compiler.
			/// </summary>
			Version^ FasmNet::GetVersion()
			{
				// Call the native function to get the version
				int nativeVersion = fasm_GetVersion();
				// Create and return a managed version object
				return gcnew Version(nativeVersion & 0xff, (nativeVersion >> 16) & 0xff);
			}
			#pragma endregion
			#pragma region FasmNet::Assemble
			/// <summary>
			/// Assembles the given mnemonics.
			/// </summary>
			/// <param name="source">The mnemonics to assemble.</param>
			/// <param name="memorySize">The memory size allocated for the buffer.</param>
			/// <param name="passLimit">The maximum number of pass to perform.</param>
			array<byte>^ FasmNet::Assemble(String^ source, int memorySize, int passLimit)
			{
				// If the string is null or empty
				if(String::IsNullOrEmpty(source))
				{
					// Return an empty array
					return gcnew array<byte>{};
				}

				// Check if the memory size was given
				if(memorySize < 1)
					memorySize = DEFAULT_MEMORY_SIZE;

				// Check if the number of maximum pass was given
				if(passLimit < 1)
					passLimit = DEFAULT_PASS_LIMIT;

				//Create some useful vars
				IntPtr nativeSource, nativeBuffer;

				try
				{
					// Allocate a chunk of memory for the native mnemonics array
					nativeSource = Marshal::StringToHGlobalAnsi(source);

					// Allocate the buffer for the assembly code
					nativeBuffer = Marshal::AllocHGlobal(memorySize);

					// Assemble the mnemonics
					FasmResults ret = (FasmResults)fasm_Assemble((char*)nativeSource.ToPointer(), (byte*)nativeBuffer.ToPointer(), memorySize, passLimit, NULL);
					NativeFasmState* state = static_cast<NativeFasmState*>(nativeBuffer.ToPointer());

					// If all was right
					if(ret == FasmResults::Ok)
					{
						// Get the assembly code
						array<byte>^ assembly = gcnew array<byte>(state->OutputLength);
						Marshal::Copy((IntPtr)state->OutputData, assembly, 0, state->OutputLength);
						return assembly;
					}
					else
					{
						throw gcnew FasmAssemblerException((FasmErrors)state->ErrorCode, state->ErrorLine->LineNumber, state->ErrorLine->FileOffset, source);
					}
				}
				finally
				{
					// Free the resources
					if(nativeSource != IntPtr::Zero)
						Marshal::FreeHGlobal(nativeSource);

					if(nativeBuffer != IntPtr::Zero)
						Marshal::FreeHGlobal(nativeBuffer);
				}
			}

			/// <summary>
			/// Assembles the given mnemonics.
			/// </summary>
			/// <param name="source">The mnemonics to assemble.</param>
			/// <remarks>The default memory size used is 4096 bytes and the maximum number of pass is 100.</remarks>
			array<byte>^ FasmNet::Assemble(String^ source)
			{
				return Assemble(source, DEFAULT_MEMORY_SIZE, DEFAULT_PASS_LIMIT);
			}

			/// <summary>
			/// Assembles the given mnemonics.
			/// </summary>
			/// <param name="source">The array containing mnemonics to assemble.</param>
			/// <param name="memorySize">The memory size allocated for the buffer.</param>
			/// <param name="passLimit">The maximum number of pass to perform.</param>
			array<byte>^ FasmNet::Assemble(array<String^>^ source, int memorySize, int passLimit)
			{
				return Assemble(String::Join(Environment::NewLine, source), memorySize, passLimit);
			}

			/// <summary>
			/// Assembles the given mnemonics.
			/// </summary>
			/// <param name="source">The array containing mnemonics to assemble.</param>
			/// <remarks>The default memory size used is 4096 bytes and the maximum number of pass is 100.</remarks>
			array<byte>^ FasmNet::Assemble(array<String^>^ source)
			{
				return Assemble(String::Join(Environment::NewLine, source), DEFAULT_MEMORY_SIZE, DEFAULT_PASS_LIMIT);
			}
			#pragma endregion
			#pragma region FasmNet::AssembleFile
			/// <summary>
			/// Assembles the specified file.
			/// </summary>
			/// <param name="path">The path of the file to assemble.</param>
			/// <param name="memorySize">The memory size allocated for the buffer.</param>
			/// <param name="passLimit">The maximum number of pass to perform.</param>
			array<byte>^ FasmNet::AssembleFile(String^ path, int memorySize, int passLimit)
			{
				return AssembleFiles(gcnew array<String^> { path }, memorySize, passLimit);
			}

			/// <summary>
			/// Assembles the specified file.
			/// </summary>
			/// <param name="path">The path of the file to assemble.</param>
			/// <param name="memorySize">The memory size allocated for the buffer.</param>
			/// <param name="passLimit">The maximum number of pass to perform.</param>
			/// <remarks>The default memory size used is 4096 bytes and the maximum number of pass is 100.</remarks>
			array<byte>^ FasmNet::AssembleFile(String^ path)
			{
				return AssembleFile(path, DEFAULT_MEMORY_SIZE, DEFAULT_PASS_LIMIT);
			}
			#pragma endregion
			#pragma region FasmNet::AssembleFiles
			/// <summary>
			/// Assembles the specified files by appending them.
			/// </summary>
			/// <param name="paths">The path of the files to assemble.</param>
			/// <param name="memorySize">The memory size allocated for the buffer.</param>
			/// <param name="passLimit">The maximum number of pass to perform.</param>
			array<byte>^ FasmNet::AssembleFiles(IEnumerable<String^>^ paths, int memorySize, int passLimit)
			{
				// Allocate a builder to store the content of all the file
				StringBuilder^ ret = gcnew StringBuilder();

				// For each file
				for each(String^ path in paths)
				{
					ret->AppendLine(File::ReadAllText(path));
				}

				// Assemble the builder
				return Assemble(ret->ToString(), memorySize, passLimit);
			}

			/// <summary>
			/// Assembles the specified files by appending them.
			/// </summary>
			/// <param name="paths">The path of the files to assemble.</param>
			/// <param name="memorySize">The memory size allocated for the buffer.</param>
			/// <param name="passLimit">The maximum number of pass to perform.</param>
			/// <remarks>The default memory size used is 4096 bytes and the maximum number of pass is 100.</remarks>
			array<byte>^ FasmNet::AssembleFiles(IEnumerable<String^>^ paths)
			{
				return AssembleFiles(paths, DEFAULT_MEMORY_SIZE, DEFAULT_PASS_LIMIT);
			}
			#pragma endregion
			#pragma endregion

			#pragma region Constructors
			/// <summary>
			/// Initializes a new instance of the <see cref="Fasm"/> class.
			/// </summary>
			/// <remarks>The default memory size used is 4096 bytes and the maximum number of pass is 100.</remarks>
			FasmNet::FasmNet()
			{
				// Call the overloaded constructor (no delegating constructor in C++/CLI :'( )
				this->FasmNet::FasmNet(DEFAULT_MEMORY_SIZE, DEFAULT_PASS_LIMIT);
			}
			/// <summary>
			/// Initializes a new instance of the <see cref="Fasm"/> class.
			/// </summary>
			/// <param name="memorySize">The memory size allocated for the buffer.</param>
			/// <param name="passLimit">The maximum number of pass to perform.</param>
			FasmNet::FasmNet(int memorySize, int passLimit)
			{
				// Save the parameters
				_memorySize = memorySize;
				_passLimit = passLimit;
				// Create a StringBuilder instance
				_mnemonics = gcnew StringBuilder();
			}
			#pragma endregion

			#pragma region Methods
			#pragma region FasmNet::AddLine
			/// <summary>
			/// Adds the text representation of the specified array of objects, followed by the current line terminator.
			/// </summary>
			/// <param name="format">The composite format string.</param>
			/// <param name="args">The array of objects to write using format.</param>
			void FasmNet::AddLine(String^ format, ... array<Object^>^ args)
			{
				_mnemonics->AppendLine(String::Format(format, args));
			}
#pragma endregion
#pragma region FasmNet::Assemble
			/// <summary>
			/// Assembles the mnemonics.
			/// </summary>
			array<byte>^ FasmNet::Assemble()
			{
				return Assemble(IntPtr::Zero);
			}
			/// <summary>
			/// Assembles the mnemonics with a given origin address.
			/// </summary>
			/// <param name="origin">The address used as starting address for the assebmly code.</param>
			array<byte>^ FasmNet::Assemble(IntPtr^ origin)
			{
				// If an origin was submitted
				if(origin != IntPtr::Zero)
					InsertLine(0, "org {0}", origin);

				// Assemble and return the assembly code
				return FasmNet::Assemble(Mnemonics, _memorySize, _passLimit);
			}
			#pragma endregion
			#pragma region FasmNet::Clear
			/// <summary>
			/// Removes all characters from the current <see cref="Fasm" /> instance.
			/// </summary>
			void FasmNet::Clear()
			{
				_mnemonics->Clear();
			}
			#pragma endregion
			#pragma region FasmNet::InsertLine
			/// <summary>
			/// Inserts the text representation of the specified array of objects, followed by the current line terminator at the specified character position.
			/// </summary>
			/// <param name="index">The position in this instance where insertion begins.</param>
			/// <param name="format">The composite format string.</param>
			/// <param name="args">The array of objects to write using format.</param>
			void FasmNet::InsertLine(int index, String^ format, ... array<Object^>^ args)
			{
				_mnemonics->Insert(index, String::Format(format, args) + Environment::NewLine);
			}
			#pragma endregion
			#pragma endregion
		}
	}
}

File: Fasm.NET.Tests/Fasm.NET.cs
/*
* FasmNet.NET
* https://github.com/ZenLulz/FasmNet.NET
*
* Copyright 2013 ZenLulz ~ Jämes Ménétrey
* Released under the MIT license
*
* Date: 2013-07-10
*/

using System;
using System.IO;
using Binarysharp.Assemblers.Fasm;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace FasmNET.Tests
{
    [TestClass]
    public class FasmNetTests
    {
        /// <summary>
        /// Gets the version of the compiler.
        /// </summary>
        [TestMethod]
        public void GetVersion()
        {
            // Act
            var version = FasmNet.GetVersion();

            // Assert
            Assert.AreEqual(1, version.Major);
            Assert.AreEqual(70, version.Minor);
        }

        /// <summary>
        /// Assembles mnemonics and checks all is right.
        /// </summary>
        [TestMethod]
        public void Assemble()
        {
            // Act
            var asm = FasmNet.Assemble("use32\nretn");

            // Assert
            Assert.AreEqual(0xc3, asm[0]);
        }

        /// <summary>
        /// Assembles wrong mnemonics and check the error.
        /// </summary>
        [TestMethod]
        public void AssembleWithError()
        {
            // Arrange
            const string mnemonics = "use32\nretnj";

            // Act
            try
            {
                FasmNet.Assemble(mnemonics);
                Assert.Fail("The above line must throw an error.");
            }
            catch (FasmAssemblerException ex)
            {
                // Assert
                Assert.AreEqual(2, ex.ErrorLine, "The error does not indicate the correct line.");
                Assert.AreEqual(6, ex.ErrorOffset, "The error does not indicate the correct offset.");
                Assert.AreEqual(FasmErrors.IllegalInstruction, ex.ErrorCode, "The kind of the error was not found.");
                Assert.AreEqual(mnemonics, ex.Mnemonics);
            }
        }

        /// <summary>
        /// Assembles 64-bit mnemonics using an array.
        /// </summary>
        [TestMethod]
        public void Assemble64BitWithArray()
        {
            // Arrange
            var mnemonics = new[]
                {
                    "use64",
                    "push rax"
                };

            // Act
            var asm = FasmNet.Assemble(mnemonics);

            // Assert
            Assert.AreEqual(0x50, asm[0]);
        }

        /// <summary>
        /// Assembles a file.
        /// </summary>
        [TestMethod]
        public void AssembleFile()
        {
            // Arrange
            const string path = "file.asm";

            // Act
            using (var file = File.CreateText(path))
            {
                file.WriteLine("use32");
                file.WriteLine("push eax");
            }

            var asm = FasmNet.AssembleFile(path);
            File.Delete(path);

            // Assert
            Assert.AreEqual(0x50, asm[0]);
        }

        /// <summary>
        /// Assembles some files.
        /// </summary>
        [TestMethod]
        public void AssembleFiles()
        {
            // Arrange
            const string path1 = "file1.asm";
            const string path2 = "file2.asm";

            // Act
            using (var file = File.CreateText(path1))
            {
                file.WriteLine("use32");
                file.WriteLine("push eax");
            }
            using (var file = File.CreateText(path2))
            {
                file.WriteLine("pop eax");
            }

            var asm = FasmNet.AssembleFiles(new[] { path1, path2 });
            File.Delete(path1);
            File.Delete(path2);

            // Assert
            CollectionAssert.AreEqual(new byte[] { 0x50, 0x58 }, asm);
        }

        /// <summary>
        /// Create an instance of the object and checks the methods AddLine, InsertLine.
        /// </summary>
        [TestMethod]
        public void ObjectAddAndInsertLines()
        {
            // Arrange
            var fasm = new FasmNet();

            // Act
            fasm.AddLine("push eax");
            fasm.AddLine("pop {0}", "eax");
            fasm.InsertLine(0, "use32");
            var asm = fasm.Assemble();

            // Assert
            CollectionAssert.AreEqual(new byte[] { 0x50, 0x58 }, asm);
        }

        /// <summary>
        /// Create an instance of the object and checks the property Mnemonics.
        /// </summary>
        [TestMethod]
        public void ObjectMnemonics()
        {
            // Arrange
            var fasm = new FasmNet();

            // Act
            fasm.AddLine("push eax");
            fasm.AddLine("retn");

            // Assert
            Assert.AreEqual(String.Format("push eax{0}retn{0}", Environment.NewLine), fasm.Mnemonics);
        }

        /// <summary>
        /// Create an instance of the object and checks the method Clear.
        /// </summary>
        [TestMethod]
        public void ObjectClear()
        {
            // Arrange
            var fasm = new FasmNet();

            // Act
            fasm.AddLine("retn");
            fasm.Clear();
            fasm.AddLine("push eax");
            fasm.AddLine("pop {0}", "eax");
            fasm.InsertLine(0, "use32");
            var asm = fasm.Assemble();

            // Assert
            CollectionAssert.AreEqual(new byte[] { 0x50, 0x58 }, asm);
        }

        /// <summary>
        /// Create an instance of the object and checks the assmble result with an origin.
        /// </summary>
        [TestMethod]
        public void ObjectAssembleWithOrigin()
        {
            // Arrange
            var fasm = new FasmNet();

            // Act
            fasm.AddLine("use32");
            fasm.AddLine("jmp {0}", 0x2000);
            var asm = fasm.Assemble(new IntPtr(0x1000));

            // Assert
            CollectionAssert.AreEqual(new byte[] { 0xE9, 0xfb, 0x0f, 0x00, 0x00 }, asm);
        }
    }
}